server/aesdsocket.c

```diff
@@ -272,33 +362,83 @@ int main(int argc, char *argv[])
 
     syslog(LOG_INFO, "aesdsocket server listening on port %s", SERVER_PORT);
 
-    // 10) Main server loop: accept connections until a signal arrives
+    // 12) Create the timer thread (for 10-second timestamps)
+    if (pthread_create(&g_timer_thread, NULL, timer_thread_func, NULL) != 0)
+    {
+        syslog(LOG_ERR, "Failed to create timer thread: %s", strerror(errno));
+        // We could continue, but no timestamps would be written
+    }
+
+    // 13) Main server loop: accept connections until g_exit_flag is set
     struct sockaddr_in client_addr;
     socklen_t addr_len = sizeof(client_addr);
 
-    while (1)
+    while (!g_exit_flag)
     {
-        g_clientfd = accept(g_socketfd, (struct sockaddr*)&client_addr, &addr_len);
-        if (g_clientfd == -1)
+        int new_fd = accept(g_socketfd, (struct sockaddr*)&client_addr, &addr_len);
+        if (new_fd == -1)
         {
-            // If accept is interrupted by a signal, break
-            if (errno == EINTR)
+            // If accept is interrupted by a signal (EINTR), check if we are shutting down
+            if (errno == EINTR && g_exit_flag)
             {
-                break;
+                break; // exit loop
             }
             syslog(LOG_ERR, "accept failed: %s", strerror(errno));
             continue;
         }
 
-        // Handle client interaction
-        serve_client(g_clientfd, &client_addr);
+        // Allocate a new list node for this client thread
+        struct thread_list_node *new_node = malloc(sizeof(*new_node));
+        if (!new_node)
+        {
+            syslog(LOG_ERR, "Failed to allocate memory for thread node");
+            close(new_fd);
+            continue;
+        }
+        memset(new_node, 0, sizeof(*new_node));
+        new_node->client_fd = new_fd;
+        memcpy(&new_node->client_addr, &client_addr, sizeof(client_addr));
+
+        // Create the client thread
+        if (pthread_create(&new_node->thread_id, NULL, client_thread_func, new_node) != 0)
+        {
+            syslog(LOG_ERR, "pthread_create failed: %s", strerror(errno));
+            close(new_fd);
+            free(new_node);
+            continue;
+        }
+
+        // Insert the new node into the singly linked list
+        SLIST_INSERT_HEAD(&g_thread_list_head, new_node, entries);
+    }
+
+    // 14) Stop accepting new connections, join existing client threads
+    struct thread_list_node *node;
+    while (!SLIST_EMPTY(&g_thread_list_head))
+    {
+        node = SLIST_FIRST(&g_thread_list_head);
+        pthread_join(node->thread_id, NULL); // Wait for the client thread to finish
```

Comment:
What's the drawback of only `pthread_join` ing outside the `while (!g_exit_flag)` loop?